### PR TITLE
Give `Span` a niche

### PR DESCRIPTION
**Pull request by Noratrieb**

### CODE REVIEW

**Comment by Nilstrieb** [comment]
Hmm, a little unsure about the regressions. Some are clearly noise like deep-vector, which regressed inside LLVM GVN). But the primary benchmarks look more legit.
`unicode-normalization`
```
1,562,628  ???:rustc_interface::callbacks::track_span_parent
  920,569  ???:<rustc_parse::lexer::StringReader>::next_token
  755,569  ???:<rustc_span::span_encoding::Span as rustc_serialize::serialize::Decodable<rustc_query_impl::on_disk_cache::CacheDecoder>>::decode
  509,229  ???:<[rustc_middle::mir::LocalDecl] as rustc_serialize::serialize::Encodable<rustc_metadata::rmeta::encoder::EncodeContext>>::encode
```
`html5ever`
```
5,880,836  ???:<rustc_expand::mbe::macro_parser::TtParser>::parse_tt::<rustc_expand::mbe::macro_rules::NoopTracker>
 -651,950  ???:<alloc::rc::Rc<alloc::vec::Vec<rustc_ast::tokenstream::TokenTree>> as core::ops::drop::Drop>::drop
  542,201  ???:<rustc_parse::lexer::StringReader>::next_token
  490,341  ???:<alloc::vec::Vec<rustc_ast::tokenstream::TokenTree> as core::ops::drop::Drop>::drop
  242,679  ???:rustc_expand::mbe::transcribe::transcribe
```

All of these functions use `Span::new`. Did that function get significantly slower because of the check. There is only one way to find out whether that really is the problem and whether it may be worth it to investigate this further and making more things use `NonMaxU32` to avoid the check. By... dropping it for a perf run. I don't think any test files have a size of 4GB^^

File: compiler/rustc_span/src/span_encoding.rs [hunk 1]
```diff
@@ -102,6 +108,8 @@ impl Span {
 
         let (base, len, ctxt2) = (lo.0, hi.0 - lo.0, ctxt.as_u32());
 
+        // FIXME: THIS IS UNSOUND. WE DO NOT CHECK FOR THE SIZE BEFORE CREATING THE NonMaxU32!!!
+
         if len <= MAX_LEN && ctxt2 <= MAX_CTXT {
```
Comment (BoxyUwU) [inline compiler/rustc_span/src/span_encoding.rs:113]: I think this is unsound, we don't check for the size before creating the nonmaxu32

**Review by BoxyUwU — CHANGES_REQUESTED** [review]

**Comment by eholk** [comment]
Marking as awaiting on author while it's a draft.

@rustbot author

**Comment by Nilstrieb** [comment]
this is a regression because of the additional branch needed. we'd need to use a niched type throughout the relevant code to try making it not a regression, which takes a little effort. I'll try to do that at some point, but I'll put this on hold for now.

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Pull request closed without merging**